### DISCUSSION TITLE
IsarError: Cannot open Environment: MdbxError (12): Out of memory

### DISCUSSION THREAD

**Discussion by ruinalsu** [isar/isar#738] [Help]
Hello,

Maybe anyone faced with following issue:
IsarError: Cannot open Environment: MdbxError (12): Out of memory
#0 nCall(package:Isar/src/native/isar_core.dart: 175)
1 openIsarSync <anonymous closure> (package:Isar/src/native/open.dart: 143)
2 using (package:ffi/src/arena.dart: 124)
3 openIsarSync (package:Isar/src/native/open.dart: 131)
4 Isar.openSync(package: Isar/src/Isar.dart: 130)

Currently reproduced **only on ZTE blade А31** (2GB RAM, 32GB storage, Android 11 (Go edition))

Tried to reproduce on Android Emulator with 128Mb RAM and 16Mb Heap - not reproduced.

It is very important for us to understand root cause.

Thanks in advance,

**Comment by Szugalew** [comment]
I've seen many of my app's users having this issue through Crashlytics.

The Crashlytics stack trace is.
```
Non-fatal Exception: io.flutter.plugins.firebase.crashlytics.FlutterError
IsarError: Cannot open Environment: MdbxError (12): Out of memory. Error thrown null.
FirebaseCrashlytics.recordError (firebase_crashlytics.dart:119)
.main (main.dart:222)
```

However, I'm not sure how to reproduce it or fix it.

I think this issue first appeared after switching to Isar 3.

**Reply by simc** [reply]
What flutter version do you use?

How many Isar instances do you open?

**Reply by Szugalew** [reply]
Flutter 3.3.1, channel stable
Dart 2.18.0

I only open one instance. I collected the effected devices from Crashyltics in case it might be helpful.

<details><summary>Devices</summary>

moto e(7i) power
248 events

Galaxy A02 (SM-A022M)
240 events

moto e(7)
239 events

moto e6 play
233 events

moto e6s
220 events

moto e⁶ plus (moto e(6) plus)
189 events

moto e5 play
155 events

moto e(6i)
144 events

Galaxy A01 Core (SM-A013M)
112 events

moto g(6)
105 events

Redmi 8A
103 events

moto e(7) power
98 events

moto g(8) play
97 events

K22 (LM-K200)
89 events

Galaxy J6+ (SM-J610G)
87 events

Redmi 6
79 events

HUAWEI Y5 2018 (DRA-LX3)
71 events

K51S (LM-K510)
70 events

Redmi 7A (Redmi 7A)
57 events

LM-K420
57 events

moto g pure
56 events

LG Q6 (LG-M700)
56 events

moto g(6) play
55 events

Moto E (5) (moto e5)
55 events

Twist 2 Pro
51 events

Redmi 6A
47 events

LG K11 (LM-X410.F)
43 events

LG K41S (LM-K410)
41 events

Moto E (4) Plus
39 events

ZTE Blade A3 2020
36 events

LG K40S (LM-X430)
36 events

Alcatel 1V (5007A)
36 events

Redmi 9C (M2006C3MG)
35 events

SMART (Infinix X657B)
35 events

SPARK Go 2021(SPARK 6 Go) (TECNO KE5)
34 events

Moto G (5S) Plus
34 events

moto g(7) play
33 events

K50S (LM-X540)
33 events

LG K10 (2017) (LG-M250)
33 events

HOT Play (Infinix X688C)
33 events

Moto G (5S)
32 events

moto e5 plus
29 events

ZTE Blade L210
28 events

Galaxy J4 Core (SM-J410G)
28 events

Moto Z (2) Play (Moto Z2 Play)
28 events

Galaxy J4+ (SM-J415G)
27 events

LG Q60 (LM-X525)
27 events

Redmi S2 (Redmi S2)
24 events

LG K20 Plus (LGMP260)
24 events

Nokia 2.4
22 events

Nokia 1.4
20 events

LM-X420
20 events

ZTE Blade L8
19 events

Nokia 2.3
19 events

Y7 (TRT-LX3)
17 events

Redmi 9C NFC (M2006C3MNG)
17 events

LG K9 (LM-X210)
16 events

Y7 (TRT-L53)
15 events

Galaxy A10s (SM-A107M)
15 events

Galaxy A02 (SM-A022F)
15 events

Moto C Plus
15 events

Galaxy J2 Pro (SM-J250M)
14 events

Moto G (5) Plus
14 events

Moto E (4)
14 events

ZTE Blade A51
13 events

Moto G4 Play (Moto G Play)
13 events

Moto G(4) Plus (Moto G (4))
13 events

LG Stylo 4 (LM-Q710.FG)
13 events

COVET PRO LITE (COVET_PRO_LITE)
13 events

(8094M)
12 events

Galaxy A13 (SM-A135M)
11 events

K51 (LM-K500)
11 events

LG Stylus3 (LG-M400)
11 events

Redmi 8A Dual
9 events

Lenovo Tab M7 (Lenovo TB-7305F)
9 events

TECNO POP 5 (TECNO BD2p)
8 events

Galaxy J8 (SM-J810M)
8 events

LG Stylo 4 (LG-Q710AL)
8 events

C5 PLUS (C5 Plus)
8 events

Galaxy J6+ (SM-J610FN)
7 events

Galaxy J7 Prime (SM-G610F)
7 events

Galaxy A6+ (SM-A605GN)
7 events

Galaxy A11 (SM-A115M)
7 events

LOGIC L61
7 events

LG K10 Power (LG-M320)
7 events

770G
7 events

moto e
6 events

ZTE Blade A7 2020
6 events

ZTE Blade A5 2019
6 events

Moto Z Play (XT1635-02)
6 events

Galaxy J6+ (SM-J610F)
6 events

Galaxy J7 Prime (SM-G610M)
6 events

Galaxy A02s (SM-A025M)
6 events

Moto G (5th Gen) (Moto G (5))
6 events

LM-X520
6 events

LG K8(2018) (LM-X210(G))
6 events

LM-X120
6 events

A7S
6 events

Alcatel 1V (5007G)
6 events

Galaxy J5(2016) (SM-J510MN)

</details>

**Comment by nEdAy** [comment]
IsarError: Cannot open Environment: MdbxError (12): Out of memory

Flutter 3.3.4 • channel stable
Tools • Dart 2.18.2 • DevTools 2.15.0

**Reply by simc** [reply]
What device does this happen on? Which Isar version do you use? How many Isar  instances do you open?

**Reply by nEdAy** [reply]
RedMi k30 pro; Isar v3.0.2  

```
class StoreService extends GetxService {
  Future<StoreService> init() async {
    final isar = await Isar.open(
        [TaskItemSchema, TaskDraftSchema, OrderTypeSchema]);
    StoreManager(isar: isar);
    return this;
  }
}

```
StoreManager is a hungry man singleton.

**Reply by simc** [reply]
@nEdAy thanks! Does it happen every time you open an instance or only sometimes?

**Reply by nEdAy** [reply]
only happened once.
I will reply here if it happens again

**Comment by ruinalsu** [comment]
In my case, after some investigation, I found that issue is reproduced when i try to create 2 instance (i have some logical separation).
When i have merged all schemas to one instance the issue have not reproduced.

**Reply by simc** [reply]
Thanks a lot then I know what the issue is. The first instance uses all available virtual memory (that is not normal RAM) and the second one has nothing left. I'll add a setting to customize this in the next version

**Reply by nEdAy** [reply]
This problem seems to happen every time with 'flutter hoy restart' for me.

**Reply by stasinato** [reply]
> This problem seems to happen every time with 'flutter hoy restart' for me.

On Android I also had this problem at every hot restart, I solved it by disabling the inspector during Isar.open

**Reply by simc** [reply]
That's very interesting. And now you don't get it anymore?

**Reply by stasinato** [reply]
So far, no. On iOS I have never had this problem anyway

**Reply by simc** [reply]
Please keep me updated if it comes back 🙂
Thanks for your help!

**Reply by Miiite** [reply]
> Thanks a lot then I know what the issue is. The first instance uses all available virtual memory (that is not normal RAM) and the second one has nothing left. I'll add a setting to customize this in the next version

We also have this issue when trying to manipulate two different instances of `Isar`.
Just so I understand correctly, is this supposed to be properly supported by the package ? Or should we assume that we should only manipulate one instance of Isar ?

**Reply by simc** [reply]
Currently I recommend to only use one instance

**Reply by JesusJoseph** [reply]
Hi all,
I am also getting below in windows desktop apps ONLY when I try to open the same application twice.
Is there any fix coming in the future releases, because for desktop applications user can open the application twice at the same time 

[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: IsarError: Cannot open Environment: MdbxError (32): The process cannot access the file because it is being used by another process.

Thanks
Joseph

**Reply by simc** [reply]
Try using 3.0.5. It should allow you to open the app multiple times :)

**Reply by JaffaKetchup** [reply]
> Thanks a lot then I know what the issue is. The first instance uses all available virtual memory (that is not normal RAM) and the second one has nothing left. I'll add a setting to customize this in the next version

@simc This still appears to be an issue. Is there any chance of this being added soon?
I've noticed you've gone quite recently, hope you're doing OK.
If you're struggling to find the time to maintain Isar, might it help if I offer a bounty to resolve this and (an)other issue(s)? If you're interested in this, please get in touch at github@jaffaketchup.dev.

**Reply by JaffaKetchup** [reply]
Also note potentially related to https://github.com/isar/isar/issues/915#issuecomment-1333775721?

**Reply by Hilbert2048** [reply]
> > Thanks a lot then I know what the issue is. The first instance uses all available virtual memory (that is not normal RAM) and the second one has nothing left. I'll add a setting to customize this in the next version
> 
> 
> 
> @simc This still appears to be an issue. Is there any chance of this being added soon?
> 
> I've noticed you've gone quite recently, hope you're doing OK.
> 
> If you're struggling to find the time to maintain Isar, might it help if I offer a bounty to resolve this and (an)other issue(s)? If you're interested in this, please get in touch at github@jaffaketchup.dev.

yes, this issue is still exist when creating multiple isar instances， it is not fixed completely.
